File: tests/src/commons-io/commons-io/2.11.0/src/test/java/commons_io/commons_io/ClassLoaderObjectInputStreamTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package commons_io.commons_io;

import org.apache.commons.io.input.ClassLoaderObjectInputStream;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectOutputStream;
import java.io.ObjectStreamClass;
import java.io.Serializable;
import java.lang.reflect.Proxy;

import static org.assertj.core.api.Assertions.assertThat;

class ClassLoaderObjectInputStreamTest {
    @Test
    void resolvesSerializableClassesWithTheConfiguredClassLoader() throws Exception {
        ObjectStreamClass descriptor = ObjectStreamClass.lookup(SerializablePayload.class);

        try (ExposedClassLoaderObjectInputStream inputStream = new ExposedClassLoaderObjectInputStream(getClass().getClassLoader())) {
            Class<?> resolvedClass = inputStream.resolveClassFor(descriptor);

            assertThat(resolvedClass).isEqualTo(SerializablePayload.class);
        }
    }

    @Test
    void fallsBackToTheParentResolutionForPrimitiveDescriptors() throws Exception {
        ObjectStreamClass descriptor = ObjectStreamClass.lookupAny(int.class);

        try (ExposedClassLoaderObjectInputStream inputStream = new ExposedClassLoaderObjectInputStream(getClass().getClassLoader())) {
            Class<?> resolvedClass = inputStream.resolveClassFor(descriptor);

            assertThat(resolvedClass).isEqualTo(int.class);
        }
    }

    @Test
    void resolvesProxyClassesWhenInterfacesCanBeProxied() throws Exception {
        try (ExposedClassLoaderObjectInputStream inputStream = new ExposedClassLoaderObjectInputStream(getClass().getClassLoader())) {
            Class<?> proxyClass = inputStream.resolveProxyClassFor(Runnable.class.getName());

            assertThat(Proxy.isProxyClass(proxyClass)).isTrue();
            assertThat(proxyClass.getInterfaces()).containsExactly(Runnable.class);
        }
    }

    @Test
    void fallsBackToTheParentProxyResolutionForNonPublicInterfacesFromAnotherLoader() throws Exception {
        ClassLoader delegatingLoader = new ClassLoader(getClass().getClassLoader()) {
        };

        try (ExposedClassLoaderObjectInputStream inputStream = new ExposedClassLoaderObjectInputStream(delegatingLoader)) {
            Class<?> proxyClass = inputStream.resolveProxyClassFor(PackagePrivateInterface.class.getName());

            assertThat(Proxy.isProxyClass(proxyClass)).isTrue();
            assertThat(proxyClass.getInterfaces()).containsExactly(PackagePrivateInterface.class);
            assertThat(proxyClass.getClassLoader()).isEqualTo(getClass().getClassLoader());
        }
    }

    private static byte[] serializationStreamHeader() throws IOException {
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        try (ObjectOutputStream objectOutputStream = new ObjectOutputStream(outputStream)) {
            objectOutputStream.flush();
        }
        return outputStream.toByteArray();
    }

    private static final class ExposedClassLoaderObjectInputStream extends ClassLoaderObjectInputStream {
        private ExposedClassLoaderObjectInputStream(ClassLoader classLoader) throws IOException {
            super(classLoader, new ByteArrayInputStream(serializationStreamHeader()));
        }

        private Class<?> resolveClassFor(ObjectStreamClass objectStreamClass) throws IOException, ClassNotFoundException {
            return super.resolveClass(objectStreamClass);
        }

        private Class<?> resolveProxyClassFor(String... interfaces) throws IOException, ClassNotFoundException {
            return super.resolveProxyClass(interfaces);
        }
    }

    interface PackagePrivateInterface {
    }

    private static final class SerializablePayload implements Serializable {
        private static final long serialVersionUID = 1L;
    }
}

File: tests/src/com.typesafe.akka/akka-actor_2.13/2.8.8/src/test/scala/com_typesafe_akka/akka_actor_2_13/ClassLoaderObjectInputStreamTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package com_typesafe_akka.akka_actor_2_13

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectOutputStream
import java.io.Serializable

import akka.util.ClassLoaderObjectInputStream
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class ClassLoaderObjectInputStreamTest {
  @Test
  def deserializesUsingProvidedClassLoader(): Unit = {
    val expected: ClassLoaderObjectInputStreamPayload = serializablePayload()
    val bytes: Array[Byte] = serialize(expected)

    val actual: ClassLoaderObjectInputStreamPayload = deserialize(bytes, getClass.getClassLoader)
      .asInstanceOf[ClassLoaderObjectInputStreamPayload]

    assertThat(actual.message).isEqualTo(expected.message)
  }

  @Test
  def fallsBackToObjectInputStreamClassResolutionWhenProvidedClassLoaderCannotLoadClass(): Unit = {
    val expected: ClassLoaderObjectInputStreamPayload = serializablePayload()
    val bytes: Array[Byte] = serialize(expected)

    val actual: ClassLoaderObjectInputStreamPayload = deserialize(bytes, null)
      .asInstanceOf[ClassLoaderObjectInputStreamPayload]

    assertThat(actual.message).isEqualTo(expected.message)
  }

  private def serializablePayload(): ClassLoaderObjectInputStreamPayload = {
    new ClassLoaderObjectInputStreamPayload("akka serialization")
  }

  private def serialize(value: AnyRef): Array[Byte] = {
    val bytes: ByteArrayOutputStream = new ByteArrayOutputStream()
    val output: ObjectOutputStream = new ObjectOutputStream(bytes)
    try {
      output.writeObject(value)
    } finally {
      output.close()
    }
    bytes.toByteArray
  }

  private def deserialize(bytes: Array[Byte], classLoader: ClassLoader): AnyRef = {
    val input: ClassLoaderObjectInputStream = new ClassLoaderObjectInputStream(
      classLoader,
      new ByteArrayInputStream(bytes))
    try {
      input.readObject().asInstanceOf[AnyRef]
    } finally {
      input.close()
    }
  }
}

final class ClassLoaderObjectInputStreamPayload(val message: String) extends Serializable
